vhost-device-sound/src/stream.rs

```diff
@@ -239,85 +244,92 @@ impl Default for PcmParams {
             .finish()
     }
 }
 
-impl Buffer {
-    pub fn new(
-        data_descriptor: virtio_queue::Descriptor,
-        message: Arc<IOMessage>,
-        direction: Direction,
-    ) -> Self {
+impl Request {
+    pub fn new(len: usize, message: Arc<IOMessage>, direction: Direction) -> Self {
         Self {
             pos: 0,
-            data_descriptor,
+            len,
             message,
             direction,
         }
     }
 
     pub fn read_output(&self, buf: &mut [u8]) -> Result<u32> {
-        let addr = self.data_descriptor.addr();
-        let offset = self.pos as u64;
-        let len = self
-            .message
-            .desc_chain
-            .memory()
-            .read(
-                buf,
-                addr.checked_add(offset)
-                    .expect("invalid guest memory address"),
-            )
-            .map_err(|_| Error::DescriptorReadFailed)?;
-        Ok(len as u32)
+        let mem = self.message.desc_chain.memory();
+
+        let Ok(mut reader) = self.message.desc_chain.clone().reader(mem) else {
+            return Err(Error::DescriptorReadFailed);
+        };
+
+        let Ok(mut reader_content) = reader.split_at(size_of::<VirtioSoundPcmXfer>() + self.pos)
+        else {
+            return Err(Error::DescriptorReadFailed);
+        };
+
+        let Ok(bytes_read) = reader_content.read(buf) else {
+            return Err(Error::DescriptorReadFailed);
+        };
+
+        Ok(bytes_read as u32)
     }
 
     pub fn write_input(&mut self, buf: &[u8]) -> Result<u32> {
-        if self.desc_len() <= self.pos as u32 {
-            return Ok(0);
-        }
-        let addr = self.data_descriptor.addr();
-        let offset = self.pos as u64;
-        let len = self
-            .message
-            .desc_chain
-            .memory()
-            .write(
-                buf,
-                addr.checked_add(offset)
-                    .expect("invalid guest memory address"),
-            )
-            .map_err(|_| Error::DescriptorWriteFailed)?;
-        self.pos += len;
-        Ok(len as u32)
+        let mem = self.message.desc_chain.memory();
+
+        let Ok(mut writer) = self.message.desc_chain.clone().writer(mem) else {
+            return Err(Error::DescriptorReadFailed);
+        };
+
+        let Ok(mut _status) = writer.split_at(self.len) else {
+            return Err(Error::DescriptorReadFailed);
+        };
+
+        let Ok(mut write_content) = writer.split_at(self.pos) else {
+            return Err(Error::DescriptorReadFailed);
+        };
+
+        let Ok(bytes_written) = write_content.write(buf) else {
+            return Err(Error::DescriptorWriteFailed);
+        };
+
+        Ok(bytes_written as u32)
     }
 
     #[inline]
     /// Returns the length of the sound data [`virtio_queue::Descriptor`].
-    pub fn desc_len(&self) -> u32 {
-        self.data_descriptor.len()
+    pub const fn len(&self) -> u32 {
+        self.len as u32
```

Comment:
Any reason this needs to return an u32? This seems like it could potentially overflow?

Reply:
I was thinking to keep it as u32 since I use it for `used_len` in https://github.com/MatiasVara/vhost-device/blob/3b97c14d9457d838c8b44d64e3c3c48b30da47c3/vhost-device-sound/src/stream.rs#L343

Reply:
I wonder if I should make `self.len` also u32. At the moment, it is usize.

Reply:
I think `used_len` must be u32 for the spec, but a request can be bigger than `used_len`.
`used_len` tracks only the bytes written by the device, so IMHO is better to use `usize` and check for overflow when we set `used_len`.

Reply:
I'll use `usize` for `self.len()` and then check for overflow when getting `used_len`.